pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/builder/FSTSerializerImpl.java

```diff
@@ -0,0 +1,370 @@
+
+  /**
+   * Supported flags.
+   */
+  private final static EnumSet<FSTFlags> FST_FLAGS = EnumSet.of(FSTFlags.NUMBERS,
+      FSTFlags.SEPARATORS, FSTFlags.FLEXIBLE, FSTFlags.STOPBIT, FSTFlags.NEXTBIT);
+
+  /**
+   * @see ImmutableFST#_filler
+   */
+  public byte _fillerByte = ImmutableFST.DEFAULT_FILLER;
+
+  /**
+   * @see ImmutableFST#_annotation
+   */
+  public byte _annotationByte = ImmutableFST.DEFAULT_ANNOTATION;
+
+  /**
+   * <code>true</code> if we should serialize with numbers.
+   *
+   * @see #withNumbers()
+   */
+  private boolean _withNumbers;
+
+  /**
+   * A hash map of [state, offset] pairs.
+   */
+  private Int2IntOpenHashMap _offsets = new Int2IntOpenHashMap();
+
+  /**
+   * A hash map of [state, right-language-count] pairs.
+   */
+  private Int2IntOpenHashMap _numbers = new Int2IntOpenHashMap();
+
+  /**
+   * A hashmap of output symbols
+   */
+  private Map<Integer, Integer> _outputSymbols = new HashMap<>();
+
+  /**
+   * Serialize the automaton with the number of right-language sequences in each
+   * node. This is required to implement perfect hashing. The numbering also
+   * preserves the order of input sequences.
+   *
+   * @return Returns the same object for easier call chaining.
+   */
+  public FSTSerializerImpl withNumbers() {
+    _withNumbers = true;
+    return this;
+  }
+
+  /**
+   * Serialize root state <code>s</code> to an output stream in
+   * <code>ImmutableFST</code> format.
+   *
+   * @see #withNumbers()
+   * @return Returns <code>os</code> for chaining.
+   */
+  @Override
+  public <T extends OutputStream> T serialize(final FST fst, T os)
+      throws IOException {
+
+    // Prepare space for arc offsets and linearize all the states.
+    int[] linearized = linearize(fst);
+
+    /*
+     * Calculate the number of bytes required for the node data, if
+     * serializing with numbers.
+     */
+    int nodeDataLength = 0;
+    if (_withNumbers) {
+      _numbers = FSTUtils.rightLanguageForAllStates(fst);
+      int maxNumber = _numbers.get(fst.getRootNode());
+      while (maxNumber > 0) {
+        nodeDataLength++;
+        maxNumber >>>= 8;
+      }
+    }
+
+    // Calculate minimal goto length.
+    int gtl = 1;
+    while (true) {
+      // First pass: calculate offsets of states.
+      if (!emitArcs(fst, null, linearized, gtl, nodeDataLength)) {
+        gtl++;
+        continue;
+      }
+
+      // Second pass: check if goto overflows anywhere.
+      if (emitArcs(fst, null, linearized, gtl, nodeDataLength)) {
+        break;
+      }
+
+      gtl++;
+    }
+
+    /*
+     * Emit the header.
+     */
+    FSTHeader.write(os, ImmutableFST.VERSION);
+    os.write(_fillerByte);
+    os.write(_annotationByte);
+    os.write((nodeDataLength << 4) | gtl);
+
+    if (IS_TRACE_ACTIVATED) {
+      System.out.println("Current buffer after emitting header and marker bytes: " + os);
+    }
+
+    DataOutputStream dataOutputStream = new DataOutputStream(os);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *dataOutputStream*. Currently, there are execution paths that do not contain closure statements, for example, when an else condition at line 189 evaluates to true. Ensure *dataOutputStream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).